File: src/rubik/solvers/tests/solvers.py
#!/usr/bin/env python3
import logging
import unittest
from rubik import shapes
from rubik.solvers import BruteForce, MeetInMiddle, MeetInMiddleRecursive
from common.cli import unit_test_parse
from common.variables import variable_names_and_objects
from common.logger import log
from common.timing import Timeout


class Solvers(unittest.TestCase):

    def setUp(self):
        self.names_and_solvers = variable_names_and_objects(BruteForce, MeetInMiddle, MeetInMiddleRecursive)

    def _check_solution_recovers_shuffled(self, *args, original, shuffled, solution_path, shuffle_path, solver_name, **kwargs):
        solution = original
        for move, reverse in zip(solution_path.moves, solution_path.reverses):
            solution = solution.move(move, reverse=reverse)
        log.debug(f"We are comparing: {solution = !s}{shuffled = !s}")
        self.assertEqual(solution, shuffled, f"Using {solver_name} we do not arrive at the target using the solution moves.")
        solution = shuffled
        for move, reverse in reversed(list(zip(solution_path.moves, solution_path.reverses))):
            other_direction = not reverse
            solution = solution.move(move, reverse=other_direction)
        self.assertEqual(solution, original, f"Using {solver_name} we do not arrive at the start using the solution moves in reverse.")

    def test_solvers(self):
        for name, Solver in self.names_and_solvers:
            solver = Solver()
            shape = shapes.Volume()
            for turns in range(4):
                with Timeout(5):
                    shuffled, shuffle_path = shape.shuffle(turns=turns, seed=0)
                    solution_path = solver.solve(start=shape, target=shuffled)
                    self.assertLessEqual(len(solution_path.moves), len(shuffle_path), f"The solution path for {name} is not optimal, as it takes more moves than we used to shuffle our cube.")
                    self._check_solution_recovers_shuffled(original=shape, shuffled=shuffled, solution_path=solution_path, shuffle_path=shuffle_path, solver_name=name)

    def test_solver_at_specific_difficulty(self):
        for name, Solver in self.names_and_solvers:
            solver = Solver()
            shape = shapes.Volume()
            from rubik.shuffles.difficulty import specific
            for turns in range(4):
                with Timeout(5):
                    shuffled, shuffle_path = specific(start=shape, turns=turns, seed=0)
                    self.assertEqual(len(shuffle_path), turns, f"We have not generated a shuffled path with the required number of turns.")
                    solution_path = solver.solve(start=shape, target=shuffled)
                    self.assertLessEqual(len(solution_path.moves), len(shuffle_path), f"The solution path for {name} is not optimal, as it takes more moves than we used to shuffle our cube.")
                    self.assertEqual(len(solution_path.moves), turns, f"Our solution path is not the right length.")
                    self._check_solution_recovers_shuffled(original=shape, shuffled=shuffled, solution_path=solution_path, shuffle_path=shuffle_path, solver_name=name)


if __name__ == '__main__':
    unit_test_parse()
    unittest.main()

File: src/rubik/solvers/tests/solvers.cpp
#include "testing/testing.h"
#include "rubik/shapes/shape.hpp"
#include "rubik/shapes/volume.hpp"
#include "rubik/shapes/cube.hpp"
#include "rubik/shufflers/shufflers.hpp"
#include "rubik/solvers/meet_in_middle_recursive.hpp"


template<typename S>
class ShapeSolverTest : public testing::Test {
};

using ShapeTypes = ::testing::Types<Volume, Cube>;
TYPED_TEST_SUITE(ShapeSolverTest, ShapeTypes);

auto check_solution_recovers_shuffled(const auto &original, const auto &shuffled, const auto &solution_path,
                                      const auto &shuffle_path) {
    auto solution = original;
    for (const auto &[move, reverse]: solution_path) {
        solution = solution.move(move, reverse);
    }
    LOG_DEBUG << "We are comparing the solution " << solution << " with the shuffled " << shuffled;
    ASSERT_EQ(solution, shuffled) << "We do not arrive at the target using the solution moves.";
    solution = shuffled;
    for (const auto &[move, reverse]: solution_path.reversed()) {
        solution = solution.move(move, reverse);
    }
    ASSERT_EQ(solution, original) << "We do not arrive at the start using the solution moves in reverse.";
}


TYPED_TEST(ShapeSolverTest, solver) {
    std::vector solvers = {MeetInMiddleRecursive<TypeParam>()};
    for (const auto &solver: solvers) {
        TypeParam shape;
        for (size_t turns = 0, seed = 0; turns < 6; turns++) {
            auto [shuffled, shuffle_path] = shape.shuffle(turns, seed);
            auto solution_path = solver.solve(shape, shuffled);
            ASSERT_LE(solution_path.size(), shuffle_path.size())
                                        << "The solution path for {solver_name} is not optimal, as it takes more moves than we used to shuffle our shape.";
            LOG_DEBUG << "The original shape is: " << shape
                      << "The shuffled shape is: " << shuffled
                      << "The shuffled shape was achieved by: " << shuffle_path
                      << "The original shape is recovered (hopefully) by: " << solution_path;
            check_solution_recovers_shuffled(shape, shuffled, solution_path, shuffle_path);
        }
    }
}

TYPED_TEST(ShapeSolverTest, specific_difficulty) {
    std::vector solvers = {MeetInMiddleRecursive<TypeParam>()};
    for (const auto &solver: solvers) {
        TypeParam shape;
        for (int turns = 0; turns < 5; turns++) {
            std::variant<int, std::string> _turns = turns;
            std::optional<int> seed = 0;
            auto [shuffled, shuffle_path] = specific(shape, _turns, seed);
            ASSERT_EQ(shuffle_path.size(), turns)
                                        << "We have not generated a shuffled path with the required number of turns.";
            auto solution_path = solver.solve(shape, shuffled);
            ASSERT_LE(solution_path.size(), shuffle_path.size())
                                        << "The solution path is not optimal, as it takes more moves than we used to shuffle our shape.";
            ASSERT_EQ(solution_path.size(), turns) << "Our solution path is not the right length.";
            check_solution_recovers_shuffled(shape, shuffled, solution_path,
                                             shuffle_path);
        }
    }
}

